.tekton/central-db-pull-request.yaml

```diff
@@ -240,10 +243,39 @@ spec:
       - name: source
         workspace: workspace
 
+    - name: determine-image-tag
+      params:
+        - name: image
+          value: $(params.output-image)
+      runAfter:
+      - clone-repository
+      taskSpec:
+        params:
+          - name: image
+            description: Image without tag
+        results:
+          - name: image-with-tag
+            description: Fully Qualified Image Name with identifying tag
+        steps:
+          - name: determine-image-tag
+            image: quay.io/stackrox-io/apollo-ci:stackrox-build-0.3.66
```

Comment:
The things I tried here:
- ❌ use a plain ubi image and install `make` & `git` - after sifting through installation errors, `go` was still missing.
- ❌ used `openshift-golang-builder` (has `make`, `git`, `go` installed): Because it is Brew based, it could not be pulled: https://github.com/stackrox/stackrox/runs/23823234740. -> might be possible to do, but the Task would need pull secrets.
- ✅ `stackrox-build` image is public, contains the necessary binaries.

Reply:
`quay.io/stackrox-io/apollo-ci` is not good because it was not produced by Konflux or downstream. Meaning that it creates (higher) supply chain security risk since it's built outside of systems that are considered safe.

In my opinion, the fact that `build-container` task "automagically" uses pull secrets but those aren't "automagically" used at the pipeline level is worth reporting as an improvement request to the Konflux team. If these are pull secrets, they should ideally work consistently in different places of the pipeline.

I see two ways to approach the supply chain security:
1. There's nothing in the `Makefile` that _requires_ `go` command for `tag` target, I believe. It's just a coincidence that some `go` commands run when the `Makefile` gets interpreted. There might be a way to refactor it so that `go` is not a requirement for `make tag`.
2. Since we have a pull secret for `brew.registry.redhat.io`, you can see a way to use it in the task.

Reply:
I went with option 1 and moved the `tag` target to a separate Makefile, which we call directly in Konflux and include in the main Makefile. I am relying on successful GHA builds & e2e tests to verify the absence of regressions. 

Using a pull secret was not possible, because:
- TaskRuns (in Tekton PipelineRuns or standalone) use a service account that has access to pull secrets
  -  there are two interesting ones in our workspace, both of which have pull and mountable secrets for each component in Konflux:
      1. `default` 
      1. `appstudio-pipeline` - in addition to above, can _mount_ our custom secrets for brew, registry.redhat.io, quay.io/rhacs-eng -> these are used in buildah task.
    - AFAICT there was no option to add the brew secret as an image pull secret to the service account (without asking Konflux team, who may or may not chose to do this). Also, we're not allowed to create new SAs in the namespace, so cannot craft one that has the required imagepullsecrets as we like.

Reply:
One follow-up (also to satisfy the EC requirements for task bundles) is to turn the inline task definition into a proper task, which is bundled and can be pulled from a trusted source. This is already captured in https://issues.redhat.com/browse/ROX-22196 .

Reply:
Regarding secrets.

Disclaimer: I prefer the approach with the `Makefile` and so the question about secrets is for my education only.

Can you please explain or point me to docs that tell what are _mountable_ secrets? I wonder how these work and under which conditions they get mounted. Asking because I could not (obviously) spot `buildah login` commands when scrolling through buildah task and for me it's total mystery how image pull secrets become available for builds. Perhaps this information will help me realize why this does not happen for pulling containers in a custom task.

----

Regarding ROX-22196.

ROX-22196 is currently an umbrella ticket for any custom task. This is fine but Ralph Bean [expressed an interest](https://redhat-internal.slack.com/archives/C04PZ7H0VA8/p1713188634789549?thread_ts=1713181360.407219&cid=C04PZ7H0VA8) to have image tag generation task contributed to <https://github.com/redhat-appstudio/build-definitions>. While we could create a task for ourselves, contributing it to community sounds really cool (although, again, not mandatory).

For this idea to not be lost, it should be captured somewhere. Right now, neither code nor ROX-22196 tells us about it. It would be nice if you create a dedicated ticket for factoring out image tag determination task and mention redhat-appstudio/build-definitions opportunity there.

Reply:
> Can you please explain or point me to docs that tell what are mountable secrets?

I think these are good explanations:
* https://livebook.manning.com/concept/kubernetes/mountable-secret#:~:text=Understanding%20a%20ServiceAccount's%20mountable%20secrets&text=By%20default%2C%20a%20pod%20can,Secrets%20on%20the%20Service%2DAccount.
* https://stackoverflow.com/a/72139678
* https://tekton.dev/docs/pipelines/auth/#basic-auth-for-docker

TLDR: mountableSecrets: An allowlist for which secrets a service account can mount and use to talk to K8s API. 

`buildah pull` (https://github.com/redhat-appstudio/build-definitions/blob/main/task/buildah/0.1/buildah.yaml#L182-L186) looks for authentication files: 

> Path of the authentication file. Default is ${XDG_RUNTIME_DIR}/containers/auth.json. See containers-auth.json(5) for more information. This file is created using buildah login.
> If the authorization state is not found there, $HOME/.docker/config.json is checked, which is set using docker login.
> From: https://github.com/containers/buildah/blob/main/docs/buildah-pull.1.md
 
I don't know where it is done, but the taskRuns of `build-container` task contain the `serviceAccountName: appstudio-pipeline` annotation. The resulting pods mount all the secrets of the service account (ie including the custom pull secrets) to a place that buildah seems to understand. They're not K8s imagePullSecrets, because the pulling of the base images takes place inside the task/pod, not on K8s level. 

---
> Regarding [ROX-22196](https://issues.redhat.com/browse/ROX-22196).

Added https://issues.redhat.com/browse/ROX-23812 as split from ROX-22196 and a comment in the code.